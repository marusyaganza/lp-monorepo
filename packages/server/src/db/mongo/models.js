const { formatIds, formatFilter } = require('../../utils/dbUtils');
const { ObjectId } = require('mongodb');

function createModel(db, table) {
  async function findOne(filter = {}) {
    const collection = await db.collection(table);
    const result = await collection.findOne(formatFilter(filter));

    return formatIds(result);
  }

  async function findMany(filter = {}) {
    const result = await db.collection(table).find(formatFilter(filter));
    const data = [];
    await result.forEach(doc => {
      data.push(formatIds(doc));
    });
    return data;
  }

  async function createOne(fields) {
    const data = { ...fields, createdAt: Date.now() };
    if (fields?.user) {
      data.user = ObjectId(fields.user);
    }
    const result = await db.collection(table).insertOne(data);

    return formatIds(result.ops[0]);
  }

  async function updateOne(fields) {
    const data = { ...fields };
    delete data.id;
    delete data.user;
    const result = await db
      .collection(table)
      .updateOne(
        { _id: ObjectId(fields.id), user: ObjectId(fields.user) },
        { $set: { ...data } }
      );
    return result?.result;
  }

  async function deleteOne(filter) {
    console.log('filter', filter);
    const result = await db.collection(table).deleteOne(formatFilter(filter));
    return result?.result;
  }

  async function createMany(entries) {
    const formattedEntries = entries.map(entry => ({
      ...entry,
      createdAt: Date.now()
    }));
    const result = await db.collection(table).insertMany(formattedEntries);
    return result.ops.map(formatIds);
  }

  return {
    findOne,
    findMany,
    createOne,
    createMany,
    updateOne,
    deleteOne
  };
}

module.exports = createModel;
