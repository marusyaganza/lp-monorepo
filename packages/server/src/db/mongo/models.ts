import { formatIds, formatFilter } from '../../utils/dbUtils';

import { Table, ModelType, DbData } from '..';

const { ObjectId } = require('mongodb');

function createModel<T extends DbData>(db: any, table: Table): ModelType<T> {
  async function findOne(filter = {}) {
    const collection = await db.collection(table);
    const result = await collection.findOne(formatFilter(filter));

    return formatIds(result);
  }

  async function findMany(filter = {}) {
    const result: T[] = await db.collection(table).find(formatFilter(filter));
    const data: T[] = [];
    await result.forEach(doc => {
      data.push(formatIds(doc));
    });
    return data;
  }

  async function createOne(fields: Partial<T>) {
    const data = { ...fields, createdAt: Date.now() };
    if (fields?.user) {
      data.user = ObjectId(fields.user);
    }
    const result = await db.collection(table).insertOne(data);

    return formatIds(result.ops[0]);
  }

  async function updateOne(fields: Partial<T>) {
    const data = { ...fields };
    delete data?.id;
    delete data?.user;
    const result = await db
      .collection(table)
      .updateOne(
        { _id: ObjectId(fields.id), user: ObjectId(fields.user) },
        { $set: { ...data } }
      );
    const isOk = result?.result?.ok ? true : false;
    const modifiedCount = result?.result?.nModified;
    return { ok: isOk, modifiedCount };
  }

  async function deleteOne(filter: Partial<T>) {
    const result = await db.collection(table).deleteOne(formatFilter(filter));
    const isOk = result?.result?.ok ? true : false;
    return { ok: isOk };
  }

  return {
    findOne,
    findMany,
    createOne,
    updateOne,
    deleteOne
  };
}

export default createModel;
