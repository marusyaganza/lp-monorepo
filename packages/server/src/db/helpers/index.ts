export * from './sortingHelpers/sortingHelpers';
export * from './general/general';
