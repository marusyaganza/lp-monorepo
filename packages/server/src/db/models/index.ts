import { UserModel, UserModelType } from './User';
import { WordModel, WordModelType } from './Word';
import { WordTagModel, WordTagModelType } from './WordTag';

export interface ModelsType {
  User: UserModelType;
  Word: WordModelType;
  WordTag: WordTagModelType;
}

export const models: ModelsType = {
  User: UserModel,
  Word: WordModel,
  WordTag: WordTagModel
};
