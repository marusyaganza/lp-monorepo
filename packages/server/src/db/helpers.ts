import { Document, Types } from 'mongoose';
import {
  Maybe,
  Language,
  Game,
  SortBy,
  SortWordsBy
} from '../generated/graphql';

type ModelDataType = {
  id?: Maybe<string>;
  user?: string;
};

interface DbData extends ModelDataType {
  _id?: string;
}

type DocumentType<T> =
  | (Document<unknown, unknown, T> &
      T & {
        _id: Types.ObjectId;
      })
  | null;

export interface WordsFilterType {
  sortBy?: SortBy | SortWordsBy | 'updatedAt';
  language: Language;
  isReverseOrder?: boolean;
  timesToLearn?: number | null;
  gameType?: Game;
  user?: string;
  tags?: string[];
  searchQuery?: string;
}

export interface WordsWithPaginationFilter extends WordsFilterType {
  page?: number;
  limit?: number;
}

export function formatData<T>(data: DocumentType<T>): T | null {
  if (!data) {
    return null;
  }
  return data.toObject({
    getters: true,
    versionKey: false
  });
}

export function formatFilter<T extends DbData>(filter: T): T {
  const id = filter?.id;
  if (id) {
    const formattedFilter = { ...filter, _id: id };
    delete formattedFilter.id;
    return formattedFilter;
  }
  return filter;
}
const NO_TAGS_ID = '000000000000000000000000';

export function getWordsFilters(filter: WordsFilterType) {
  const {
    user,
    language = Language.English,
    sortBy,
    isReverseOrder,
    gameType,
    tags,
    searchQuery,
    timesToLearn = 5
  } = filter;

  // filter for a specific game
  let gameFilter;
  if (gameType === Game.Conjugation) {
    gameFilter = { conjugation: { $ne: null } };
  }
  if (gameType === Game.Audio) {
    gameFilter = { audioUrl: { $ne: '' } };
  }

  // filter based on tags
  let tagsFilters: { tags: string }[] = [];
  if (tags?.length) {
    tagsFilters = tags
      ?.filter(tag => tag !== NO_TAGS_ID)
      .map(tag => {
        return { tags: tag };
      });
  }

  let noTagsFilter;
  if (tags?.includes(NO_TAGS_ID)) {
    noTagsFilter = { $or: [{ tags: [] }, { tags: null }] };
  }

  let searchFilter;
  if (searchQuery) {
    const searchRegexp = { $regex: new RegExp(searchQuery, 'i') };
    searchFilter = {
      $or: [
        { name: searchRegexp },
        { 'defs.def': searchRegexp },
        { alternativeSpelling: searchRegexp },
        { stems: searchRegexp }
      ]
    };
  }

  // filter based on learning status
  let learningStatusFilters: Record<string, unknown>[] = [];
  if (gameType && sortBy !== SortBy.MemoryRefresher) {
    learningStatusFilters = [
      { isLearned: { $ne: true } },
      {
        $or: [
          { [`statistics.${gameType}.successRate`]: { $lt: timesToLearn } },
          { [`statistics.${gameType}.successRate`]: null }
        ]
      }
    ];
  }
  if (sortBy === SortBy.MemoryRefresher) {
    learningStatusFilters = [
      {
        $or: [
          { [`statistics.${gameType}.successRate`]: { $gte: timesToLearn } },
          { isLearned: true }
        ]
      }
    ];
  }

  //sorting
  let orderNum = isReverseOrder ? -1 : 1;

  // words with the highest number of errors should go first
  if (sortBy === SortBy.ErrorCount) {
    orderNum = -1 * orderNum;
  }

  let sort: Record<string, number> = { $natural: orderNum };

  if (sortBy) {
    let propName: string = sortBy;
    if (gameType) {
      propName = `statistics.${gameType}.${sortBy}`;
    }

    if (sortBy !== SortBy.MemoryRefresher) {
      sort = { [propName]: orderNum };
    }
  }

  const filtersArray = [
    gameFilter,
    noTagsFilter,
    searchFilter,
    ...tagsFilters,
    ...learningStatusFilters
  ].filter(Boolean);

  const filters: Record<string, unknown> = {
    user,
    language
  };

  if (filtersArray?.length) {
    filters.$and = filtersArray;
  }

  return {
    sort,
    filters
  };
}
