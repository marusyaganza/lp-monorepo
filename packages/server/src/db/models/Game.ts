import { GameConfig, Language } from '../../generated/graphql';
import { formatData, formatFilter } from '../helpers';
import { Game } from '../schema/Game';
// import { games as gamesMock } from '../../mocks/games';

export interface GameFilterType extends Partial<GameConfig> {
  languages?: Language;
}

export interface GameModelType {
  findOne: (filter: GameFilterType) => Promise<GameConfig | null>;
  findMany: (filter: GameFilterType) => Promise<GameConfig[] | null>;
}

export const GameModel: GameModelType = {
  async findOne(filter) {
    const game = await Game.findOne(formatFilter(filter));
    return formatData(game);
  },

  async findMany(filter) {
    const games = await Game.find(formatFilter(filter));
    // await Game.insertMany(gamesMock);
    return games;
  }
};
