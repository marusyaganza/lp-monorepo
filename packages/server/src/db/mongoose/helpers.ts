import { Types, Document } from 'mongoose';
import { User, Word } from '../../generated/graphql';
import { DbData } from 'db';

type DocumentType =
  | (Document<unknown, Record<string, any>, User | Word> &
      Omit<
        | User
        | (Word & {
            _id: Types.ObjectId;
          }),
        never
      >)
  | null;

export function formatData(data: DocumentType) {
  if (!data) {
    return null;
  }
  return data.toObject({
    getters: true,
    versionKey: false
  });
}

export function formatFilter<T extends DbData>(filter: T): T {
  const id = filter.id;
  if (id) {
    const formattedFilter = { ...filter, _id: id };
    delete formattedFilter.id;
    return formattedFilter;
  }
  return filter;
}
