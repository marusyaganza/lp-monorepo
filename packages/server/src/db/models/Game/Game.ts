import {
  GameConfig,
  Language,
  Game as GameType
} from '../../../generated/graphql';
import { Game } from '../../schema/Game';

export interface GameFilterType {
  languages?: Language;
  type?: GameType;
}

export interface GameModelType {
  findOne: (filter: GameFilterType) => Promise<GameConfig | null | undefined>;
  findMany: (
    filter: GameFilterType
  ) => Promise<GameConfig[] | null | undefined>;
}

export const GameModel: GameModelType = {
  async findOne(filter) {
    const game = await Game.findOne(filter);
    return game?.toObject();
  },

  async findMany(filter) {
    const games = await Game.find(filter).sort('orderNum').exec();
    return games;
  }
};
