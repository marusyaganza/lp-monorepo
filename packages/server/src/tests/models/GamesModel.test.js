import { GameModel } from '../../db/models/Game';
import { Game, Language } from '../../generated/graphql';
import { connectToDb, disconnectFromDb, dropDb, seedDb } from '../helpers';

const snapshotConfig = {
  id: expect.any(String),
  _id: expect.any(Object)
};

describe('WordTagModel findMany', () => {
  beforeEach(async () => {
    await connectToDb();
    await dropDb();
    await seedDb();
  });

  afterEach(async () => {
    await dropDb();
    await disconnectFromDb();
  });

  test('English', async () => {
    const result = await GameModel.findMany({
      languages: Language.English
    });
    expect(result.length).toBe(4);
  });
  test('Spanish', async () => {
    const result = await GameModel.findMany({
      languages: Language.Spanish
    });

    expect(result.length).toBe(5);
  });
});

const gameTypes = Object.values(Game);

describe('WordTagModel findOne', () => {
  beforeEach(async () => {
    await connectToDb();
    await seedDb();
  });

  afterEach(async () => {
    await dropDb();
    await disconnectFromDb();
  });
  gameTypes.forEach(game => {
    test(`with ${game} game, English`, async () => {
      const result = await GameModel.findOne({
        type: game
      });
      expect(result).toMatchSnapshot(snapshotConfig);
    });
  });
});
