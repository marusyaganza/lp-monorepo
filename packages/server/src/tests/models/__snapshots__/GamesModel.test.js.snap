// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordTagModel findOne with AUDIO game, English 1`] = `
{
  "_id": Any<Object>,
  "desc": "Listen, Type, Master. Sharpen language skills through auditory engagement.",
  "id": Any<String>,
  "languages": [
    "ENGLISH",
    "SPANISH",
  ],
  "minWords": 1,
  "name": "WordEcho",
  "optionsPerGame": 0,
  "orderNum": 0,
  "timesToLearn": 5,
  "type": "AUDIO",
  "wordsPerGame": 6,
}
`;

exports[`WordTagModel findOne with CONJUGATION game, English 1`] = `
{
  "_id": Any<Object>,
  "desc": "Conjugate verbs!",
  "id": Any<String>,
  "languages": [
    "SPANISH",
  ],
  "minWords": 1,
  "name": "Conjugation",
  "optionsPerGame": 0,
  "orderNum": 4,
  "timesToLearn": 5,
  "type": "CONJUGATION",
  "wordsPerGame": 2,
}
`;

exports[`WordTagModel findOne with SELECT_DEF game, English 1`] = `
{
  "_id": Any<Object>,
  "desc": "Define & Conquer.Match words with their meanings intuitively.",
  "id": Any<String>,
  "languages": [
    "ENGLISH",
    "SPANISH",
  ],
  "minWords": 1,
  "name": "LexiSense",
  "optionsPerGame": 6,
  "orderNum": 1,
  "timesToLearn": 5,
  "type": "SELECT_DEF",
  "wordsPerGame": 6,
}
`;

exports[`WordTagModel findOne with SELECT_WORD game, English 1`] = `
{
  "_id": Any<Object>,
  "desc": "Associate definitions with the right words. A captivating word puzzle adventure!",
  "id": Any<String>,
  "languages": [
    "ENGLISH",
    "SPANISH",
  ],
  "minWords": 1,
  "name": "Definify: Match & Learn",
  "optionsPerGame": 6,
  "orderNum": 2,
  "timesToLearn": 5,
  "type": "SELECT_WORD",
  "wordsPerGame": 6,
}
`;

exports[`WordTagModel findOne with TYPE_WORD game, English 1`] = `
{
  "_id": Any<Object>,
  "desc": "Type words matching given definitions. Boost vocabulary in engaging rounds!",
  "id": Any<String>,
  "languages": [
    "ENGLISH",
    "SPANISH",
  ],
  "minWords": 1,
  "name": "WordRecall: Define & Type",
  "optionsPerGame": 0,
  "orderNum": 3,
  "timesToLearn": 5,
  "type": "TYPE_WORD",
  "wordsPerGame": 6,
}
`;
